lib/improver/convection.py

```diff
@@ -0,0 +1,197 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing convection diagnosis utilities."""
+
+import iris
+import numpy as np
+
+from improver.utilities.spatial import DifferenceBetweenAdjacentGridSquares
+from improver.threshold import BasicThreshold
+from improver.nbhood import NeighbourhoodProcessing
+
+
+class DiagnoseConvectivePrecipitation(object):
+    """
+    Diagnose convective precipitation by using sharp differences between
+    adjacent grid squares to help distinguish convective precipitation with
+    sharp features from stratiform precipitation with broader (less sharp)
+    features.
+    """
+
+    def __init__(
+            self, lower_threshold, higher_threshold, neighbourhood_method,
+            radii_in_km, fuzzy_factor=None, below_thresh_ok=False,
+            lead_times=None, unweighted_mode=False, ens_factor=1.0,
+            use_adjacent_grid_square_differences=True):
+        """
+        Args:
+            lower_threshold : float
+                The threshold point for 'significant' datapoints to define the
+                lower threshold e.g. 0 mm/hr.
+            higher_threshold : float
+                The threshold point for 'significant' datapoints to define the
+                higher threshold e.g. 5 mm/hr.
+            fuzzy_factor : float or None
+                Percentage above or below threshold for fuzzy membership value.
+                If None, no fuzzy_factor is applied.
+            neighbourhood_method : str
+                Name of the neighbourhood method to use. Options: 'circular'.
+            radii_in_km : float or List (if defining lead times)
```

Comment:
Additionally, all the "_in_km" stuff should be ditched. We are moving to metres as the default and that change will be checked in before this one. Unit tests etc will need updating where appropriate as well.